adapters/rubicon/rubicon.go

```
@@ -941,6 +909,104 @@ func resolveBidFloor(bidFloor float64, bidFloorCur string, reqInfo *adapters.Ext
+		}
+		if len(app.PageCat) > 0 {
+			target = addStringArrayAttribute(app.PageCat, target, "pagecat")
+		}
+	}
+
+	impExtContextAttributes, _, _, _ := jsonparser.Get(imp.Ext, "context", "data")
+	impExtDataAttributes, _, _, _ := jsonparser.Get(imp.Ext, "data")
+
+	if impExtContextAttributes != nil && len(impExtContextAttributes) > 0 {
+		target = populateFirstPartyDataAttributes(impExtContextAttributes, target)
+	} else if impExtDataAttributes != nil && len(impExtDataAttributes) > 0 {
+		target = populateFirstPartyDataAttributes(impExtDataAttributes, target)
+	}
+
+	if len(extImp.Keywords) > 0 {
+		target = addStringArrayAttribute(extImp.Keywords, target, "keywords")
+	}
+
+	return target, nil
+}
+
+func addStringAttribute(attribute string, target json.RawMessage, attributeName string) json.RawMessage {
+	targetAsMap := messageToMap(target)
+	targetAsMap[attributeName] = [1]string{attribute}
+
+	updatedTarget, _ := json.Marshal(targetAsMap)
+	return updatedTarget
+}
+
+func addStringArrayAttribute(attribute []string, target json.RawMessage, attributeName string) json.RawMessage {
+	targetAsMap := messageToMap(target)
+	targetAsMap[attributeName] = attribute
+
+	updatedTarget, _ := json.Marshal(targetAsMap)
+	return updatedTarget
+}
+
+func getAdSlot(imp openrtb2.Imp) string {
+	adServerContextName, _ := jsonparser.GetString(imp.Ext, "context", "data", "adserver", "name")
+
+	if adServerContextName == "gam" {
+		contextAdSlot, _ := jsonparser.GetString(imp.Ext, "context", "data", "adserver", "adslot")
+		return contextAdSlot
+	}
+	adServerDataName, _ := jsonparser.GetString(imp.Ext, "data", "adserver", "name")
+	if adServerDataName == "gam" {
+		dataAdSlot, _ := jsonparser.GetString(imp.Ext, "data", "adserver", "adslot")
+		return dataAdSlot
+	}
+
+	return ""
+}
+
+func updateUserRpTargetWithFpdAttributes(visitor json.RawMessage, user openrtb2.User) (json.RawMessage, error) {
+	existingTarget, _, _, _ := jsonparser.Get(user.Ext, "rp", "target")
```

Please do not ignore errors and add appropriate error handling logic here

---

If I will not have this entry I will just have existingTarget as nil, and error `Key path not found`, which can be skipped by existing logic

---

Please add error handling logic for errors other than key path not found for the same reason as above.